### DISCUSSION TITLE
Editor integration

### DISCUSSION THREAD

**Discussion by Gregoor** [softwaremill/walk-that-type#7] [Ideas]
Hi! Good stuff, thanks for putting it out there.

I'm curious to hear your thoughts on possible editor integrations and whether that is an avenue that you are thinking about yourself?

The types that I need to debug draw in enough dependencies that debugging them in a browser text area is not feasible, and I'd love to be able to use these powers straight from VSCode. Happy to help with an extension for that one as well, I've put a few together myself in the past.

Best,
Gregor

**Comment by mieszkosabo** [comment]
Hi @Gregoor thank you for your suggestion! 

Having a VSCode integration is definitely something I had in mind since the beginning as well. It would be lovely if one could click on a type like `type _sth = ComplexType<"hello">;` and I guess some panel would show up with the UI for stepping through it. At least, that's one idea, I'm 100% open for other propositions.

As far as implementation goes... I could transform the repo into a monorepo, so that the extension source code could live alongside the rest of the source code and import React components or just the utility functions from the interpreter engine. That's pretty easy.

What I'm worried the most is getting the correct environment (as in type definitions) for the Walk That Type. Currently you paste/write all needed type definitions in the embedded editor in the app and WTT internally creates a single typescript file with them that is passed to the TS compiler in order to get the AST (abstract syntax tree), which is then mapped into an internal representation.

I guess for every type that is used inside the definition of `ComplexType` one could use the language server to "go to definition", recursively do the same for all types used in it, and then return an array of definitions. With all definitions present, we could proceed as in the web app, by mapping them into internal representations.

But this is only my speculation, I haven't done anything like this before. But if you (or anyone) are willing to give it a try and confirm that something like this would be possible then I'd be very happy to co-work on implementing this feature!

**Reply by Gregoor** [reply]
Heya, thanks for the detailed reply!

I was also thinking along similar lines wrt how the UI should present itself. The flow could also start when hovering over a type, with the last few (or first?) steps of the expansion, where clicking in opens a panel. Anyway, stuff for the future.

I appreciate lining out where the difficulty lies, I'll look at that when I find prototyping time again.
